### ISSUE TITLE
`<For/>` does not react to changes tracked via parent `Store`

### ISSUE THREAD

**Issue by hWorblehat** [leptos-rs/leptos#3401]
**Describe the bug**
When using the `<For/>` component on an iterable field of a `Store`, it only reacts when the field is updated via its own signal. If the field is updated via its parent signal, no reaction occurs.

**Leptos Dependencies**

```toml
console_error_panic_hook = "0.1.7"
leptos = { version = "0.7.2", features = ["csr"] }
reactive_stores = "0.1.2"
```

**To Reproduce**

1. Run the following code as `main.rs`:
   ```rs
    use leptos::{mount::mount_to_body, prelude::*};
    use reactive_stores::Store;
    
    #[derive(Debug, Default, Clone, Store)]
    struct MyStore {
        #[store(key: u16 = |item| *item)]
        items: Vec<u16>,
        next_id: u16,
    }
    
    impl MyStore {
        fn new_item(&mut self) {
            let id = self.next_id;
            self.next_id += 1;
            self.items.push(id)
        }
    }
    
    #[component]
    fn App() -> impl IntoView {
        let store = Store::new(MyStore::default());
    
        let new_item = move || {
            let next_id = store.next_id().get();
            *store.next_id().write() = next_id + 1;
            store.items().write().push(next_id)
        };
    
        view! {
            <button on:click=move |_| {
                store.write().new_item();
                // new_item();
            }>Add</button>
            <h1>List</h1>
            <ul>
                <For
                    each=move || store.items()
                    key=|item| item.get()
                    let:item
                >
                    <li>{move || item.get()}</li>
                </For>
            </ul>
            <h1>Debug</h1>
            <p>{move ||format!("{:?}", *store.items().read())}</p>
        }
    }
    
    fn main() {
        console_error_panic_hook::set_once();
        mount_to_body(App)
    }
   ```
3. In the rendered webpage, click 'Add'
4. Notice that that the 'Debug' paragraph reacts to display the new contents of the `items` field, but the 'List' list does not.
5. _This can be "fixed" by commenting out the `store.write().new_item()` line in the button handler, and replacing it with a call to the `new_item()` closure (which mutates the `items` field directly via its `Write` implementation, rather than the parent store's)._

**Expected behavior**
The ordered list generated by the `<For/>` component should update whenever the `items` Vec is updated, regardless of whether this is done via `store.write()` or `store.items().write()`.

**Comment by hWorblehat** [comment]
This issue can also be reproduced with the [Stores example in the repository](https://github.com/leptos-rs/leptos/blob/28af33d5114a17ca8b0436a57020cdb2f26d3a6f/examples/stores/src/lib.rs) by changing line 103 from:
```rs
store.todos().write().push(Todo::new(input_ref.get().unwrap().value()));
```
to:
```rs
store.write().todos.push(Todo::new(input_ref.get().unwrap().value()));
```

**Comment by gbj** [comment]
Thanks for the issue and for the nice reproduction! Really pleasant to work with. Should be closed by #3403, but let me know if that does not fix the problem in whatever non-trivial case you're working with.
